DB/src/main/java/io/deephaven/db/v2/locations/parquet/local/ParquetColumnLocation.java

```diff
@@ -0,0 +1,332 @@
+                             final ChunkSource.GetContext firstContext = firstColumn.makeGetContext(CHUNK_SIZE);
+                             final ChunkSource.GetContext lastContext = lastColumn.makeGetContext(CHUNK_SIZE);
+                             final OrderedKeys rows = OrderedKeys.forRange(0, numRows - 1);
+                             final OrderedKeys.Iterator rowsIterator = rows.getOrderedKeysIterator()) {
+
+                            while (rowsIterator.hasMore()) {
+                                final OrderedKeys chunkRows = rowsIterator.getNextOrderedKeysWithLength(CHUNK_SIZE);
+
+                                buildGrouping.build(boxerKernel.box(keyColumn.getChunk(keyContext, chunkRows)),
+                                        firstColumn.getChunk(firstContext, chunkRows),
+                                        lastColumn.getChunk(lastContext, chunkRows));
+                            }
+
+                            metaData = buildGrouping.getGrouping();
+                        }
+                    }
+                }
+            }
+            return metaData;
+        }
+
+        private interface BuildGrouping extends Context {
+            void build(@NotNull ObjectChunk<?, ? extends Values> keyChunk,
+                       @NotNull Chunk<? extends Attributes.UnorderedKeyIndices> firstChunk,
+                       @NotNull Chunk<? extends Attributes.UnorderedKeyIndices> lastChunk);
+
+            Object getGrouping();
+
+            static BuildGrouping builder(@NotNull final ChunkType chunkType, final int numRows) {
+                switch (chunkType) {
+                    case Int:
+                        return new IntBuildGrouping(numRows);
+                    case Long:
+                        return new LongBuildGrouping(numRows);
+                    default:
+                        throw new IllegalArgumentException("Unknown type for an index: " + chunkType);
+                }
+            }
+
+            final class IntBuildGrouping implements BuildGrouping {
+
+                private final Map<Object, int[]> grouping;
+
+                IntBuildGrouping(final int numRows) {
+                    grouping = new LinkedHashMap<>(numRows);
+                }
+
+                @Override
+                public void build(@NotNull final ObjectChunk<?, ? extends Values> keyChunk,
+                                  @NotNull final Chunk<? extends Attributes.UnorderedKeyIndices> firstChunk,
+                                  @NotNull final Chunk<? extends Attributes.UnorderedKeyIndices> lastChunk) {
+                    final IntChunk<? extends Attributes.UnorderedKeyIndices> firstIntChunk = firstChunk.asIntChunk();
+                    final IntChunk<? extends Attributes.UnorderedKeyIndices> lastIntChunk = lastChunk.asIntChunk();
+
+                    for (int ki = 0; ki < keyChunk.size(); ++ki) {
+                        final int[] range = new int[2];
```

Comment:
This whole business with the naked Object `Object getGrouping()` method is pretty confusing.
The fact that we have parallel chunks, convert to wasteful two-element arrays, and then, many years and many lives later, we end up calling ParallelDeferredGroupingProvider.validateAndTransformMetadata with whatever we got from here /and that uses the data as if it was coming from parallel arrays/ is pretty maddening.

Very likely tho, not for doing anything about it now.

Reply:
No. We definitely don't want to touch this until we incorporate grouping-as-table changes from DHE.